File: frontend/lib/api/inventory.ts
import { apiClient, unwrap } from "./client";
import type {
  AdjustInventoryItemInput,
  InventoryItem,
  InventoryItemInput,
  InventoryTransaction,
  ListInventoryItemsParams,
  Paginated,
} from "@/types";

/** GET /api/v1/inventory/items/?category=&search= */
export async function listInventoryItems(
  params: ListInventoryItemsParams = {}
): Promise<Paginated<InventoryItem>> {
  const res = await apiClient.get("/inventory/items/", { params });
  return unwrap<Paginated<InventoryItem>>(res);
}

/** POST /api/v1/inventory/items/ */
export async function createInventoryItem(input: InventoryItemInput): Promise<InventoryItem> {
  const res = await apiClient.post("/inventory/items/", input);
  return unwrap<InventoryItem>(res);
}

/** GET /api/v1/inventory/items/low-stock/ */
export async function listLowStockItems(): Promise<Paginated<InventoryItem>> {
  const res = await apiClient.get("/inventory/items/low-stock/");
  return unwrap<Paginated<InventoryItem>>(res);
}

/** GET /api/v1/inventory/items/{id}/ */
export async function getInventoryItem(itemId: number): Promise<InventoryItem> {
  const res = await apiClient.get(`/inventory/items/${itemId}/`);
  return unwrap<InventoryItem>(res);
}

/** PATCH /api/v1/inventory/items/{id}/ — note `quantity` isn't settable here;
 * see adjustInventoryItem. */
export async function updateInventoryItem(
  itemId: number,
  input: Partial<InventoryItemInput>
): Promise<InventoryItem> {
  const res = await apiClient.patch(`/inventory/items/${itemId}/`, input);
  return unwrap<InventoryItem>(res);
}

/** DELETE /api/v1/inventory/items/{id}/ */
export async function deleteInventoryItem(itemId: number): Promise<void> {
  await apiClient.delete(`/inventory/items/${itemId}/`);
}

/** POST /api/v1/inventory/items/{id}/adjust/ — restock, log usage, or
 * correct a stocktake discrepancy. Always logged as an InventoryTransaction
 * (never edits `quantity` directly), so crossing the low-stock threshold
 * reliably triggers the staff notification pipeline. */
export async function adjustInventoryItem(
  itemId: number,
  input: AdjustInventoryItemInput
): Promise<InventoryItem> {
  const res = await apiClient.post(`/inventory/items/${itemId}/adjust/`, input);
  return unwrap<InventoryItem>(res);
}

/** GET /api/v1/inventory/items/{id}/transactions/ */
export async function listInventoryItemTransactions(itemId: number): Promise<Paginated<InventoryTransaction>> {
  const res = await apiClient.get(`/inventory/items/${itemId}/transactions/`);
  return unwrap<Paginated<InventoryTransaction>>(res);
}

File: Backend/apps/inventory/api/views/inventory.py
from rest_framework import generics, permissions
from rest_framework.generics import get_object_or_404
from rest_framework.response import Response
from rest_framework.views import APIView

from apps.inventory.api.serializers import (
    InventoryAdjustSerializer,
    InventoryItemSerializer,
    InventoryTransactionSerializer,
)
from apps.inventory.models import InventoryItem, InventoryTransaction


class InventoryItemListCreateView(generics.ListCreateAPIView):
    """GET/POST /api/v1/inventory/items/"""

    serializer_class = InventoryItemSerializer
    permission_classes = [permissions.IsAdminUser]

    def get_queryset(self):
        qs = InventoryItem.objects.all()
        category = self.request.query_params.get("category")
        search = self.request.query_params.get("search")
        if category:
            qs = qs.filter(category=category)
        if search:
            qs = qs.filter(name__icontains=search)
        return qs


class InventoryItemDetailView(generics.RetrieveUpdateDestroyAPIView):
    """GET/PATCH/DELETE /api/v1/inventory/items/{id}/"""

    serializer_class = InventoryItemSerializer
    permission_classes = [permissions.IsAdminUser]
    queryset = InventoryItem.objects.all()


class InventoryLowStockListView(generics.ListAPIView):
    """GET /api/v1/inventory/items/low-stock/"""

    serializer_class = InventoryItemSerializer
    permission_classes = [permissions.IsAdminUser]

    def get_queryset(self):
        from django.db.models import F

        return InventoryItem.objects.filter(quantity__lte=F("low_stock_threshold"))


class InventoryItemAdjustView(APIView):
    """POST /api/v1/inventory/items/{id}/adjust/ — restock, log usage, or
    correct a stocktake discrepancy. Always logged as an InventoryTransaction
    (never edits `quantity` directly) so there's a full audit trail."""

    permission_classes = [permissions.IsAdminUser]

    def post(self, request, pk):
        item = get_object_or_404(InventoryItem, pk=pk)
        serializer = InventoryAdjustSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        data = serializer.validated_data

        InventoryTransaction.objects.create(
            item=item,
            change_type=data["change_type"],
            quantity_change=data["quantity_change"],
            reason=data["reason"],
            created_by=request.user,
        )
        item.refresh_from_db()
        return Response(InventoryItemSerializer(item).data)


class InventoryItemTransactionListView(generics.ListAPIView):
    """GET /api/v1/inventory/items/{id}/transactions/"""

    serializer_class = InventoryTransactionSerializer
    permission_classes = [permissions.IsAdminUser]

    def get_queryset(self):
        return InventoryTransaction.objects.filter(item_id=self.kwargs["pk"])
